### DISCUSSION TITLE
JavetProxyConverter and memory management

### DISCUSSION THREAD

**Discussion by Stechmuck** [caoccao/Javet#142] [Q&A] [answered]
I've played around a bit with Javet and noticed some memory notifications.

Here is a test to reproduce:

```Java
public class JavetTest {
  @Test
  void testVar() {
    for (String script : new String[] {
        "var b = a;",
        "var b = a; b = undefined;"}) {
      System.out.println("Execute script: " + script);
      try (V8Runtime v8Runtime = V8Host.getV8Instance().createV8Runtime()) {
        v8Runtime.setConverter(new JavetProxyConverter());
        v8Runtime.getGlobalObject().set("a", new A());
        v8Runtime.getExecutor(script).executeVoid();
        v8Runtime.getGlobalObject().delete("a");
        v8Runtime.lowMemoryNotification();
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
  }

  public class A {}
}
```
And the result is:
```
Execute script: var b = a;
Feb 03, 2022 2:57:14 PM com.caoccao.javet.utils.JavetDefaultLogger warn
WARNUNG: 4 V8 callback context object(s) not recycled.
Execute script: var b = a; b = undefined;
```

So do I have to set all with a proxyobject assigned variables in Javascript to undefined before the script ends, or is there a way around this?

**Comment by caoccao** [comment, marked as answer]
The root cause of the _potential_ memory leak is: global objects are not subject to GC.

The solution is as follows.

- Use `let`, `const` instead of `var`.
- Wrap the variable declaration in a function.

Why _potential_? Javet keeps track of every memory leak point and releases all of them when V8 runtime is being closed.

**Reply by Stechmuck** [reply]
Thank you for your answer.
Wrapping into a function works.
